wordpress-mobile/gutenberg-mobile
# Correct Color for RichText links in Dark Mode

**Fixes** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2399
**Related PRs:**
`gutenberg` https://github.com/WordPress/gutenberg/pull/23289
`AztecEditor-iOS` https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1295
~`WordPress-Android` https://github.com/wordpress-mobile/WordPress-Android/pull/12226~
~`WordPress-iOS` https://github.com/wordpress-mobile/WordPress-iOS/pull/14351~

## Solution Description
As it turns out this error seems to be one that will randomly happen. Android uses the following process to set the text and style on RichText components:
- The [text is set](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java#L206) 
- This then updates the text [on AztexText](https://github.com/wordpress-mobile/gutenberg-mobile/blob/74f2bb2c58b1fcd9a10125b78673318d5ef8d0a4/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java#L221) 
- The links are [then styled](https://github.com/wordpress-mobile/AztecEditor-Android/blob/develop/aztec/src/main/kotlin/org/wordpress/aztec/AztecText.kt#L1191) as part of that call
- Which is then dependent on `linkFormatter` to [determine the span and set the style](https://github.com/wordpress-mobile/AztecEditor-Android/blob/a1386beed8d8215fe8ab79c501c1ec237ae8a785/aztec/src/main/kotlin/org/wordpress/aztec/AztecText.kt#L1374) This needs both the text and for `linkFormatter` to be set correctly.


**For the link color**
- When the `linkTextColor` [is set](https://github.com/wordpress-mobile/gutenberg-mobile/blob/74f2bb2c58b1fcd9a10125b78673318d5ef8d0a4/react-native-aztec/android/src/main/java/org/wordpress/mobile/ReactNativeAztec/ReactAztecManager.java#L391) this updates `linkFormatter` with the proper attributes. 

**The Issue**
It appears that the `@ReactProp`'s don't enforce an order of operations[[1]](https://stackoverflow.com/questions/50135103/is-there-a-way-to-enforce-order-of-execution-of-props-on-a-native-component#). So if `linkTextColor` is processed before `text` everything will work fine. If not then the app will fall back to the default styles.

**The Fix**
To fix this I simply moved the `linkTextColor` to be part of the `text` attribute that is set so that we can enforce the order of operations. 

This is actually only needed for Android however to prevent diverging implementations I made the same changes to iOS. 

## To Test
- Set your device too dark mode
- Create a post/page or open one that has links. 

Color should be using `blue-30` (`#5198d9`)

| Before | After |
| --- | --- |
|<kbr><img width=300 src="https://user-images.githubusercontent.com/3384451/85074845-38d0cd00-b18b-11ea-8502-2162fa6400cc.jpg"/></kbr> | <kbr><img width=300 src="https://user-images.githubusercontent.com/3384451/85075934-49824280-b18d-11ea-9c16-8702ecd93248.jpg"/></kbr> |

## PR submission checklist:
